### PR TITLE
Command line tool for generating kernels from JSON descriptions

### PR DESCRIPTION
**Pull request by p-otto**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,7 @@ option(OPTION_SELF_CONTAINED "Create a self-contained install with all dependenc
 option(OPTION_BUILD_TESTS    "Build tests."                                           ON)
 option(OPTION_BUILD_DOCS     "Build documentation."                                   OFF)
 option(OPTION_BUILD_EXAMPLES "Build examples."                                        OFF)
+option(OPTION_BUILD_TOOLS    "Build tools."                                           OFF)
```
Comment (scheibel) [inline CMakeLists.txt]: Tools should get build by default.

File: source/tools/glkernel-cmd/KernelGeneration.cpp [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    }
+
+    template <typename T>
+    void processCommands(glkernel::tkernel<T> & kernel, const cppexpose::VariantArray & commandArray)
+    {
+        for (auto& command : commandArray)
+        {
+            auto commandMap = command.asMap();
+
+            auto& element = *commandMap->begin();
```
Comment (scheibel) [inline source/tools/glkernel-cmd/KernelGeneration.cpp]: If the map is empty, the defererence operation is malformed on most compilers. You should check against the end pointer first.

Comment (scheibel) [inline source/tools/glkernel-cmd/KernelGeneration.cpp]: Or at least, add an assertion that commandMap isn't empty.

File: source/tools/glkernel-cmd/main.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#include <cppexpose/variant/Variant.h>
+#include <cppexpose/json/JSON.h>
+
+int main(int argc, char* argv[])
+{
+    cppassist::ArgumentParser argParser;
+    argParser.parse(argc, argv);
+
+    auto inFilename = argParser.value("--i");
+    auto outFilename = argParser.value("--o");
```
Comment (scheibel) [inline source/tools/glkernel-cmd/main.cpp]: A typical extension for command line tools is the integration in pipe architectures. We may specify an interface for this later.

File: source/tools/glkernel-cmd/main.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    std::ofstream outStream(outFilename);
+
+    if (!outStream.is_open())
+    {
+        std::cerr << "ERROR: Output file could not be created. Aborting..." << std::endl;
+        return 1;
+    }
+
+    outStream << cppexpose::JSON::stringify(kernelJSON, cppexpose::JSON::OutputMode::Beautify) << std::endl;
+}
```
Comment (scheibel) [inline source/tools/glkernel-cmd/main.cpp]: Add an explicit ```return 0;```. There should be a warning configured by cmake-init which catches this.

**Comment by p-otto** [comment]
Please review again

File: source/tools/glkernel-cmd/KernelGeneration.cpp [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+#include <glkernel/sort.h>
+
+#include <cppexpose/json/JSON.h>
+
+namespace
+{
+    template <typename T>
+    void processPoissonSquare(glkernel::tkernel<T> & kernel, unsigned int num_probes)
+    {
+        std::cout << "Poisson square is not implemented for the specified kernel type. Only kernel2 is supported." << std::endl;
```
Comment (scheibel) [inline source/tools/glkernel-cmd/KernelGeneration.cpp]: As you already have cppassist as a depdency, I suggest you use the cppassist logging interfaces for outputs:
* `cppassist::critical`
* `cppassist::error`
* `cppassist::warning`
* `cppassist::info`
* `cppassist::debug`

File: source/tools/glkernel-cmd/main.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <cppassist/cmdline/ArgumentParser.h>
+
+#include <cppexpose/variant/Variant.h>
+#include <cppexpose/json/JSON.h>
+
+int main(int argc, char* argv[])
+{
+    cppassist::ArgumentParser argParser;
+    argParser.parse(argc, argv);
```
Comment (scheibel) [inline source/tools/glkernel-cmd/main.cpp]: You should add the output of a help text if no parameter or the `--help` parameter is passed. `ArgumentParser::print()` may be a good starting point for this.

Comment (sbusch42) [inline source/tools/glkernel-cmd/main.cpp]: We already have a more sophisticated framework for parsing command line parameters in cppassist: CommandLineProgram. I suggest to use that for real command line programs like this one, it helps with specifying the possible options and handles such things as outputting help texts automatically. There is also an example for this is cppassist.

**Pull request merged**